plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -337,23 +344,43 @@ private function get_final_frontend_query( $query, $page = 1, $is_exclude_applie
 	 * @param bool  $is_exclude_applied_filters Whether to exclude the applied filters or not.
 	 */
 	private function get_final_query_args( $common_query_values, $query, $is_exclude_applied_filters = false ) {
-		$handpicked_products = $query['handpicked_products'] ?? [];
-		$orderby_query       = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : [];
+		$handpicked_products = $query['handpicked_products'] ?? array();
+		$orderby_query       = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : array();
 		$on_sale_query       = $this->get_on_sale_products_query( $query['on_sale'] );
 		$stock_query         = $this->get_stock_status_query( $query['stock_status'] );
-		$visibility_query    = is_array( $query['stock_status'] ) ? $this->get_product_visibility_query( $stock_query, $query['stock_status'] ) : [];
+		$visibility_query    = is_array( $query['stock_status'] ) ? $this->get_product_visibility_query( $stock_query, $query['stock_status'] ) : array();
 		$featured_query      = $this->get_featured_query( $query['featured'] ?? false );
 		$attributes_query    = $this->get_product_attributes_query( $query['product_attributes'] );
-		$taxonomies_query    = $query['taxonomies_query'] ?? [];
+		$taxonomies_query    = $query['taxonomies_query'] ?? array();
 		$tax_query           = $this->merge_tax_queries( $visibility_query, $attributes_query, $taxonomies_query, $featured_query );
-		$date_query          = $this->get_date_query( $query['timeFrame'] ?? [] );
+		$date_query          = $this->get_date_query( $query['timeFrame'] ?? array() );
 
 		// We exclude applied filters to generate product ids for the filter blocks.
-		$applied_filters_query = $is_exclude_applied_filters ? [] : $this->get_queries_by_applied_filters();
+		$applied_filters_query = $is_exclude_applied_filters ? array() : $this->get_queries_by_applied_filters();
 
 		$merged_query = $this->merge_queries( $common_query_values, $orderby_query, $on_sale_query, $stock_query, $tax_query, $applied_filters_query, $date_query );
 
-		return $this->filter_query_to_only_include_ids( $merged_query, $handpicked_products );
+		$result = $this->filter_query_to_only_include_ids( $merged_query, $handpicked_products );
+
+		if ( isset( $query['priceRange'] ) ) {
+			/**
+			 * We are adding these extra query arguments to be used in `posts_clauses`
+			 * because there are 2 special edge cases we wanna handle for Price range filter:
+			 * Case 1: Prices excluding tax are displayed including tax
+			 * Case 2: Prices including tax are displayed excluding tax
+			 *
+			 * Both of these cases require us to modify SQL query to get the correct results.
+			 */
+			$result = array_merge(
+				$result,
+				array(
+					'isProductCollection' => true,
+					'priceRange'          => $query['priceRange'],
+				)
+			);
+		}
```

Comment:
I'm a bit concerned about this logic here:
- Because we're now using `priceRange` and `isProductCollection` as query args, should we make them valid query arg in `get_valid_query_vars()`?
- Or, the `isProductCollection` argument is only used with `priceRange`, should we change the `priceRange` arg to something more unique so we can just pass that arg only?

After solving the issue above, I think we can pass the priceRange to the `merge_queries()` method, having to use `array_merge` with `merge_queries` make me a bit confused at the first glance.

Reply:
Totally see your points! 🙌🏻

> Or, the `isProductCollection` argument is only used with `priceRange`, should we change the `priceRange` arg to something more unique so we can just pass that arg only?
> 

I think it's a good call to stick with `isProductCollection`, as it lets us insert a check [here](https://github.com/woocommerce/woocommerce/blob/97037d847fb7b56a5cc8c85d0b4738b31d7dec7d/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1079-L1082). We could make the `priceRange` argument name more unique, but clarity wins in this case. This approach also paves the way for more intricate logic down the road. We can examine the `isProductCollection` argument and tweak the WP_Query accordingly. But hey, I'm open to ideas. If you feel renaming `priceRange` is the way to go, let's explore that. 🙂

> having to use `array_merge` with `merge_queries` make me a bit confused at the first glance.
> 

To address this, I have refactored the code in https://github.com/woocommerce/woocommerce/pull/42858/commits/97037d847fb7b56a5cc8c85d0b4738b31d7dec7d. Could you please review the changes and share your feedback? Thanks 🙏🏻